// @grpc/grpc-js-xds/build/src/generated/envoy/config/core/v3/HealthStatus.js
"use strict";
// Original file: deps/envoy-api/envoy/config/core/v3/health_check.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.HealthStatus = void 0;
/**
 * Endpoint health status.
 */
exports.HealthStatus = {
    /**
     * The health status is not known. This is interpreted by Envoy as ``HEALTHY``.
     */
    UNKNOWN: 'UNKNOWN',
    /**
     * Healthy.
     */
    HEALTHY: 'HEALTHY',
    /**
     * Unhealthy.
     */
    UNHEALTHY: 'UNHEALTHY',
    /**
     * Connection draining in progress. E.g.,
     * `<https://aws.amazon.com/blogs/aws/elb-connection-draining-remove-instances-from-service-with-care/>`_
     * or
     * `<https://cloud.google.com/compute/docs/load-balancing/enabling-connection-draining>`_.
     * This is interpreted by Envoy as ``UNHEALTHY``.
     */
    DRAINING: 'DRAINING',
    /**
     * Health check timed out. This is part of HDS and is interpreted by Envoy as
     * ``UNHEALTHY``.
     */
    TIMEOUT: 'TIMEOUT',
    /**
     * Degraded.
     */
    DEGRADED: 'DEGRADED',
};
