// @grpc/grpc-js-xds/build/src/generated/envoy/config/cluster/v3/Cluster.js
"use strict";
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetSelector_LbSubsetSelectorFallbackPolicy = exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetMetadataFallbackPolicy = exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetFallbackPolicy = exports._envoy_config_cluster_v3_Cluster_LbPolicy = exports._envoy_config_cluster_v3_Cluster_RingHashLbConfig_HashFunction = exports._envoy_config_cluster_v3_Cluster_DnsLookupFamily = exports._envoy_config_cluster_v3_Cluster_DiscoveryType = exports._envoy_config_cluster_v3_Cluster_ClusterProtocolSelection = void 0;
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
exports._envoy_config_cluster_v3_Cluster_ClusterProtocolSelection = {
    /**
     * Cluster can only operate on one of the possible upstream protocols (HTTP1.1, HTTP2).
     * If :ref:`http2_protocol_options <envoy_v3_api_field_config.cluster.v3.Cluster.http2_protocol_options>` are
     * present, HTTP2 will be used, otherwise HTTP1.1 will be used.
     */
    USE_CONFIGURED_PROTOCOL: 'USE_CONFIGURED_PROTOCOL',
    /**
     * Use HTTP1.1 or HTTP2, depending on which one is used on the downstream connection.
     */
    USE_DOWNSTREAM_PROTOCOL: 'USE_DOWNSTREAM_PROTOCOL',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * Refer to :ref:`service discovery type <arch_overview_service_discovery_types>`
 * for an explanation on each type.
 */
exports._envoy_config_cluster_v3_Cluster_DiscoveryType = {
    /**
     * Refer to the :ref:`static discovery type<arch_overview_service_discovery_types_static>`
     * for an explanation.
     */
    STATIC: 'STATIC',
    /**
     * Refer to the :ref:`strict DNS discovery
     * type<arch_overview_service_discovery_types_strict_dns>`
     * for an explanation.
     */
    STRICT_DNS: 'STRICT_DNS',
    /**
     * Refer to the :ref:`logical DNS discovery
     * type<arch_overview_service_discovery_types_logical_dns>`
     * for an explanation.
     */
    LOGICAL_DNS: 'LOGICAL_DNS',
    /**
     * Refer to the :ref:`service discovery type<arch_overview_service_discovery_types_eds>`
     * for an explanation.
     */
    EDS: 'EDS',
    /**
     * Refer to the :ref:`original destination discovery
     * type<arch_overview_service_discovery_types_original_destination>`
     * for an explanation.
     */
    ORIGINAL_DST: 'ORIGINAL_DST',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * When V4_ONLY is selected, the DNS resolver will only perform a lookup for
 * addresses in the IPv4 family. If V6_ONLY is selected, the DNS resolver will
 * only perform a lookup for addresses in the IPv6 family. If AUTO is
 * specified, the DNS resolver will first perform a lookup for addresses in
 * the IPv6 family and fallback to a lookup for addresses in the IPv4 family.
 * This is semantically equivalent to a non-existent V6_PREFERRED option.
 * AUTO is a legacy name that is more opaque than
 * necessary and will be deprecated in favor of V6_PREFERRED in a future major version of the API.
 * If V4_PREFERRED is specified, the DNS resolver will first perform a lookup for addresses in the
 * IPv4 family and fallback to a lookup for addresses in the IPv6 family. i.e., the callback
 * target will only get v6 addresses if there were NO v4 addresses to return.
 * If ALL is specified, the DNS resolver will perform a lookup for both IPv4 and IPv6 families,
 * and return all resolved addresses. When this is used, Happy Eyeballs will be enabled for
 * upstream connections. Refer to :ref:`Happy Eyeballs Support <arch_overview_happy_eyeballs>`
 * for more information.
 * For cluster types other than
 * :ref:`STRICT_DNS<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.STRICT_DNS>` and
 * :ref:`LOGICAL_DNS<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.LOGICAL_DNS>`,
 * this setting is
 * ignored.
 * [#next-major-version: deprecate AUTO in favor of a V6_PREFERRED option.]
 */
exports._envoy_config_cluster_v3_Cluster_DnsLookupFamily = {
    AUTO: 'AUTO',
    V4_ONLY: 'V4_ONLY',
    V6_ONLY: 'V6_ONLY',
    V4_PREFERRED: 'V4_PREFERRED',
    ALL: 'ALL',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * The hash function used to hash hosts onto the ketama ring.
 */
exports._envoy_config_cluster_v3_Cluster_RingHashLbConfig_HashFunction = {
    /**
     * Use `xxHash <https://github.com/Cyan4973/xxHash>`_, this is the default hash function.
     */
    XX_HASH: 'XX_HASH',
    /**
     * Use `MurmurHash2 <https://sites.google.com/site/murmurhash/>`_, this is compatible with
     * std:hash<string> in GNU libstdc++ 3.4.20 or above. This is typically the case when compiled
     * on Linux and not macOS.
     */
    MURMUR_HASH_2: 'MURMUR_HASH_2',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * Refer to :ref:`load balancer type <arch_overview_load_balancing_types>` architecture
 * overview section for information on each type.
 */
exports._envoy_config_cluster_v3_Cluster_LbPolicy = {
    /**
     * Refer to the :ref:`round robin load balancing
     * policy<arch_overview_load_balancing_types_round_robin>`
     * for an explanation.
     */
    ROUND_ROBIN: 'ROUND_ROBIN',
    /**
     * Refer to the :ref:`least request load balancing
     * policy<arch_overview_load_balancing_types_least_request>`
     * for an explanation.
     */
    LEAST_REQUEST: 'LEAST_REQUEST',
    /**
     * Refer to the :ref:`ring hash load balancing
     * policy<arch_overview_load_balancing_types_ring_hash>`
     * for an explanation.
     */
    RING_HASH: 'RING_HASH',
    /**
     * Refer to the :ref:`random load balancing
     * policy<arch_overview_load_balancing_types_random>`
     * for an explanation.
     */
    RANDOM: 'RANDOM',
    /**
     * Refer to the :ref:`Maglev load balancing policy<arch_overview_load_balancing_types_maglev>`
     * for an explanation.
     */
    MAGLEV: 'MAGLEV',
    /**
     * This load balancer type must be specified if the configured cluster provides a cluster
     * specific load balancer. Consult the configured cluster's documentation for whether to set
     * this option or not.
     */
    CLUSTER_PROVIDED: 'CLUSTER_PROVIDED',
    /**
     * Use the new :ref:`load_balancing_policy
     * <envoy_v3_api_field_config.cluster.v3.Cluster.load_balancing_policy>` field to determine the LB policy.
     * This has been deprecated in favor of using the :ref:`load_balancing_policy
     * <envoy_v3_api_field_config.cluster.v3.Cluster.load_balancing_policy>` field without
     * setting any value in :ref:`lb_policy<envoy_v3_api_field_config.cluster.v3.Cluster.lb_policy>`.
     */
    LOAD_BALANCING_POLICY_CONFIG: 'LOAD_BALANCING_POLICY_CONFIG',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * If NO_FALLBACK is selected, a result
 * equivalent to no healthy hosts is reported. If ANY_ENDPOINT is selected,
 * any cluster endpoint may be returned (subject to policy, health checks,
 * etc). If DEFAULT_SUBSET is selected, load balancing is performed over the
 * endpoints matching the values from the default_subset field.
 */
exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetFallbackPolicy = {
    NO_FALLBACK: 'NO_FALLBACK',
    ANY_ENDPOINT: 'ANY_ENDPOINT',
    DEFAULT_SUBSET: 'DEFAULT_SUBSET',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetMetadataFallbackPolicy = {
    /**
     * No fallback. Route metadata will be used as-is.
     */
    METADATA_NO_FALLBACK: 'METADATA_NO_FALLBACK',
    /**
     * A special metadata key ``fallback_list`` will be used to provide variants of metadata to try.
     * Value of ``fallback_list`` key has to be a list. Every list element has to be a struct - it will
     * be merged with route metadata, overriding keys that appear in both places.
     * ``fallback_list`` entries will be used in order until a host is found.
     *
     * ``fallback_list`` key itself is removed from metadata before subset load balancing is performed.
     *
     * Example:
     *
     * for metadata:
     *
     * .. code-block:: yaml
     *
     * version: 1.0
     * fallback_list:
     * - version: 2.0
     * hardware: c64
     * - hardware: c32
     * - version: 3.0
     *
     * at first, metadata:
     *
     * .. code-block:: json
     *
     * {"version": "2.0", "hardware": "c64"}
     *
     * will be used for load balancing. If no host is found, metadata:
     *
     * .. code-block:: json
     *
     * {"version": "1.0", "hardware": "c32"}
     *
     * is next to try. If it still results in no host, finally metadata:
     *
     * .. code-block:: json
     *
     * {"version": "3.0"}
     *
     * is used.
     */
    FALLBACK_LIST: 'FALLBACK_LIST',
};
// Original file: deps/envoy-api/envoy/config/cluster/v3/cluster.proto
/**
 * Allows to override top level fallback policy per selector.
 */
exports._envoy_config_cluster_v3_Cluster_LbSubsetConfig_LbSubsetSelector_LbSubsetSelectorFallbackPolicy = {
    /**
     * If NOT_DEFINED top level config fallback policy is used instead.
     */
    NOT_DEFINED: 'NOT_DEFINED',
    /**
     * If NO_FALLBACK is selected, a result equivalent to no healthy hosts is reported.
     */
    NO_FALLBACK: 'NO_FALLBACK',
    /**
     * If ANY_ENDPOINT is selected, any cluster endpoint may be returned
     * (subject to policy, health checks, etc).
     */
    ANY_ENDPOINT: 'ANY_ENDPOINT',
    /**
     * If DEFAULT_SUBSET is selected, load balancing is performed over the
     * endpoints matching the values from the default_subset field.
     */
    DEFAULT_SUBSET: 'DEFAULT_SUBSET',
    /**
     * If KEYS_SUBSET is selected, subset selector matching is performed again with metadata
     * keys reduced to
     * :ref:`fallback_keys_subset<envoy_v3_api_field_config.cluster.v3.Cluster.LbSubsetConfig.LbSubsetSelector.fallback_keys_subset>`.
     * It allows for a fallback to a different, less specific selector if some of the keys of
     * the selector are considered optional.
     */
    KEYS_SUBSET: 'KEYS_SUBSET',
};
